### PR TITLE
Improve startup failure

### PR DESCRIPTION
**Pull request by Donnerbart**
https://hivemq.kanbanize.com/ctrl_board/22/cards/20399/details/

### CODE REVIEW

File: src/main/java/com/hivemq/extensions/prometheus/PrometheusExtensionMain.java [hunk 1]
```diff
@@ -61,15 +64,22 @@ public void extensionStart(
...
         }
-
-        prometheusServer = new PrometheusServer(configuration, Services.metricRegistry());
-        prometheusServer.start();
+        try {
+            final PrometheusServer prometheusServer = new PrometheusServer(configuration, Services.metricRegistry());
+            prometheusServer.start();
+            prometheusServerRef.set(prometheusServer);
+        } catch (final Exception e) {
+            LOG.error("Error starting the Jetty Server for Prometheus Extension", e);
```
Comment (Donnerbart) [inline src/main/java/com/hivemq/extensions/prometheus/PrometheusExtensionMain.java:72]: We now catch all startup failures of the Jetty server here, log the original exception and prevent the extension startup properly.

File: src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java [hunk 2]
```diff
@@ -60,45 +64,48 @@ public PrometheusServer(
...
         server.setHandler(context);
         context.addServlet(new ServletHolder(new MonitoredMetricServlet(metricRegistry)), configuration.metricPath());
-        try {
-            server.start();
-        } catch (final Exception e) {
-            log.error("Error starting the Jetty Server for Prometheus Extension");
-            log.debug("Original exception was:", e);
-        }
-        log.info("Started Jetty Server exposing Prometheus Servlet on URI {}",
-                trimTrailingSlash(server.getURI().toString()) + configuration.metricPath());
```
Comment (Donnerbart) [inline src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java:85]: `server.getURI().toString()` produced the NPE when the server couldn't start.

File: src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java [hunk 3]
```diff
@@ -60,45 +64,48 @@ public PrometheusServer(
...
+        dropwizardExportsRef.set(dropwizardExports);
 
         final ServletContextHandler context = new ServletContextHandler();
         context.setContextPath("/");
         server.setHandler(context);
         context.addServlet(new ServletHolder(new MonitoredMetricServlet(metricRegistry)), configuration.metricPath());
-        try {
-            server.start();
-        } catch (final Exception e) {
-            log.error("Error starting the Jetty Server for Prometheus Extension");
```
Comment (Donnerbart) [inline src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java:81]: We only logged startup errors, but then proceed without properly preventing the extension startup. We now just throw all exceptions into the main method and handle them there.

File: src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java [hunk 4]
```diff
@@ -60,45 +64,48 @@ public PrometheusServer(
...
+            LOG.error("Exception occurred while stopping the Prometheus Extension");
+            LOG.debug("Original exception was: ", e);
         }
     }
 
-    private @NotNull String trimTrailingSlash(final @NotNull String serverUri) {
-        if (serverUri.endsWith("/")) {
-            return serverUri.substring(0, serverUri.length() - 1);
+    private @NotNull String trimTrailingSlash(final @Nullable URI serverUri) {
+        if (serverUri == null) {
```
Comment (Donnerbart) [inline src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java:102]: We should not get here, but I still made it a bit more resilient (just used for a log message).

**Review by YannickWeber — APPROVED** [review]

File: src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java [hunk 5]
```diff
@@ -44,12 +47,13 @@ public class PrometheusServer {
     /* Maximum thread count for Jetty's thread pool */
     private static final int MAX_THREADS = 8;
 
-    private static final @NotNull Logger log = LoggerFactory.getLogger(PrometheusServer.class);
+    private static final @NotNull Logger LOG = LoggerFactory.getLogger(PrometheusServer.class);
+
+    private final @Nullable AtomicReference<DropwizardExports> dropwizardExportsRef = new AtomicReference<>();
```
Comment (tlopesPT) [inline src/main/java/com/hivemq/extensions/prometheus/export/PrometheusServer.java:52]:
```suggestion
    private final @NotNull AtomicReference<DropwizardExports> dropwizardExportsRef = new AtomicReference<>();
```

**Review by tlopesPT — APPROVED** [review]

**Pull request merged**